### PR TITLE
Added a script to import git repository to ClickHouse

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added a script to import git repository to ClickHouse.

Detailed description / Documentation draft:
Better implementation of #12577

Allows to answer questions like:
- list files with maximum number of authors;
- show me the oldest lines of code in the repository;
- show me the files with longest history;
- list favorite files for author;
- list largest files with lowest number of authors;
- at what weekday the code has highest chance to stay in repository;
- the distribution of code age across repository;
- files sorted by average code age;
- quickly show file with blame info (rough);
- commits and lines of code distribution by time; by weekday, by author; for specific subdirectories;
- show history for every subdirectory, file, line of file, the number of changes (lines and commits) across time; how the number of contributors was changed across time;
- list files with most modifications;
- list files that were rewritten most number of time or by most of authors;
- what is percentage of code removal by other authors, across authors;
- the matrix of authors that shows what authors tends to rewrite another authors code;
- what is the worst time to write code in sense that the code has highest chance to be rewritten;
- the average time before code will be rewritten and the median (half-life of code decay);
- comments/code percentage change in time / by author / by location;
- who tend to write more tests / cpp code / comments.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Now the tool can easily process ClickHouse, LLVM, Linux and Chromium repositories.

ClickHouse repository processing takes 31 seconds.
[LLVM](https://github.com/llvm/llvm-project) - 8 minutes 45 seconds.
[Linux](https://github.com/torvalds/linux) - 12 minutes 35 seconds.
[Chromium](https://github.com/chromium/chromium) - 67 minutes 3 seconds.

(I picked them as an examples of somewhat large repositories. What is the largest repository on GitHub?).

The time of processing is in fact comparable what time takes to clone the repository.

**Comment by alexey-milovidov** [comment]
```
SELECT 
    database,
    table,
    sum(rows),
    formatReadableSize(sum(bytes))
FROM system.parts
WHERE database LIKE 'git\\_%'
GROUP BY 
    database,
    table
ORDER BY 
    database ASC,
    table ASC

┌─database───────┬─table────────┬─sum(rows)─┬─formatReadableSize(sum(bytes))─┐
│ git_chromium   │ commits      │    920847 │ 88.93 MiB                      │
│ git_chromium   │ file_changes │   8280114 │ 274.98 MiB                     │
│ git_chromium   │ line_changes │ 941256784 │ 18.65 GiB                      │
│ git_clickhouse │ commits      │     31900 │ 2.44 MiB                       │
│ git_clickhouse │ file_changes │    135270 │ 4.87 MiB                       │
│ git_clickhouse │ line_changes │   3300441 │ 62.27 MiB                      │
│ git_linux      │ commits      │    875169 │ 74.33 MiB                      │
│ git_linux      │ file_changes │   2088558 │ 116.73 MiB                     │
│ git_linux      │ line_changes │  85152422 │ 1.71 GiB                       │
│ git_llvm       │ commits      │    363055 │ 37.06 MiB                      │
│ git_llvm       │ file_changes │   1468936 │ 67.69 MiB                      │
│ git_llvm       │ line_changes │  62253479 │ 1.20 GiB                       │
└────────────────┴──────────────┴───────────┴────────────────────────────────┘
```

**Comment by alexey-milovidov** [comment]
> Integration tests (release) — fail: 0, passed: 786, error: 2 

```
name = 'connect', self = <socket._socketobject object at 0x7f44784b8980>
args = (('172.19.0.3', 9000),)

    def meth(name,self,*args):
>       return getattr(self._sock,name)(*args)
E       error: [Errno 113] No route to host
```

Infrastructure failure.

**Comment by alexey-milovidov** [comment]
https://github.com/ClickHouse/ClickHouse/pull/14553

**Review by KochetovNicolai — APPROVED** [review]

File: programs/git-import/git-import.cpp [hunk 1]
```diff
@@ -0,0 +1,1234 @@
...
+    std::string prev_commit_hash;
+    std::string prev_author;
+    LocalDateTime prev_time{};
+
+    /** Classify line to empty / code / comment / single punctuation char.
+      * Very rough and mostly suitable for our C++ style.
+      */
+    void setLineInfo(std::string full_line)
+    {
+        indent = 0;
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply programs/git-import/git-import.cpp]: Ok.

**Pull request merged**